File: src/main/java/fun/bookish/vertx/activemq/client/core/ActiveMQClientImpl.java
package fun.bookish.vertx.activemq.client.core;

import fun.bookish.vertx.activemq.client.cache.ActiveMQCacheManager;
import fun.bookish.vertx.activemq.client.config.ActiveMQOptions;
import fun.bookish.vertx.activemq.client.consumer.ActiveMQConsumer;
import fun.bookish.vertx.activemq.client.consumer.ActiveMQConsumerImpl;
import fun.bookish.vertx.activemq.client.pool.ActiveMQSessionPool;
import fun.bookish.vertx.activemq.client.producer.ActiveMQProducer;
import fun.bookish.vertx.activemq.client.producer.ActiveMQProducerImpl;
import fun.bookish.vertx.activemq.client.subscriber.ActiveMQSubscriber;
import fun.bookish.vertx.activemq.client.subscriber.ActiveMQSubscriberImpl;
import io.vertx.core.AsyncResult;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;


public class ActiveMQClientImpl implements ActiveMQClient {

    private final Vertx vertx;
    private final ActiveMQSessionPool sessionPool;
    private final ActiveMQCacheManager cacheManager;


    ActiveMQClientImpl(Vertx vertx,ActiveMQOptions options){
        this.vertx = vertx;
        this.sessionPool = new ActiveMQSessionPool(options);
        this.cacheManager = new ActiveMQCacheManager();
        this.vertx.getOrCreateContext().put(ActiveMQClientImpl.class.getTypeName(),this);
    }

    ActiveMQClientImpl(Vertx vertx,ActiveMQOptions options, Handler<AsyncResult<Void>> handler){
        this.vertx = vertx;
        this.sessionPool = new ActiveMQSessionPool(options, handler);
        this.cacheManager = new ActiveMQCacheManager();
        this.vertx.getOrCreateContext().put(ActiveMQClientImpl.class.getTypeName(),this);
    }

    @Override
    public ActiveMQConsumer createConsumer(String key, String destination) {
        return new ActiveMQConsumerImpl(key,vertx,sessionPool.getSession(),destination,cacheManager);
    }

    @Override
    public ActiveMQSubscriber createSubscriber(String key, String destination) {
        return new ActiveMQSubscriberImpl(key,vertx,sessionPool.getSession(),destination,cacheManager);
    }

    @Override
    public ActiveMQProducer createProducer(String key, DestinationType destinationType, String destination) {
        return new ActiveMQProducerImpl(key,vertx,sessionPool.getSession(),destinationType,destination,cacheManager);
    }

    @Override
    public ActiveMQConsumer getConsumer(String key) {
        return this.cacheManager.getConsumer(key);
    }

    @Override
    public ActiveMQSubscriber getSubscriber(String key) {
        return this.cacheManager.getSubscriber(key);
    }

    @Override
    public ActiveMQProducer getProducer(String key) {
        return this.cacheManager.getProducer(key);
    }

    @Override
    public void clear(String keyPrefix) {
        this.cacheManager.clear(keyPrefix);
    }

}

File: src/main/kotlin/cn/booklish/vertx/activemq/client/core/ActiveMQClientImpl.kt
package cn.booklish.vertx.activemq.client.core

import cn.booklish.vertx.activemq.client.consumer.ActiveMQConsumer
import cn.booklish.vertx.activemq.client.consumer.ActiveMQConsumerImpl
import cn.booklish.vertx.activemq.client.pool.ActiveMQSessionPool
import cn.booklish.vertx.activemq.client.producer.ActiveMQProducer
import cn.booklish.vertx.activemq.client.producer.ActiveMQProducerImpl
import cn.booklish.vertx.activemq.client.subscriber.ActiveMQSubscriber
import cn.booklish.vertx.activemq.client.subscriber.ActiveMQSubscriberImpl
import io.vertx.core.Vertx
import io.vertx.core.json.JsonObject
import org.apache.activemq.ActiveMQConnectionFactory
import javax.jms.Connection


class ActiveMQClientImpl(private val vertx: Vertx, config: JsonObject):ActiveMQClient{

    private val connection: Connection = ActiveMQConnectionFactory(config.getString("username"),
                                            config.getString("password"),config.getString("brokerURL"))
                    .createConnection()

    private val sessionPool = ActiveMQSessionPool(connection,config.getInteger("sessionPoolSize")?:0)

    init {
        connection.clientID = "vertx-activemq-client"
        connection.start()
    }

    override fun createConsumer(key: String,destination: String): ActiveMQConsumer {
        return ActiveMQConsumerImpl(key,vertx,sessionPool.getSession(),destination)
    }

    override fun createSubscriber(key: String,destination: String): ActiveMQSubscriber {
        return ActiveMQSubscriberImpl(key,vertx,sessionPool.getSession(),destination)
    }

    override fun createProducer(key: String,destinationType: DestinationType,destination: String): ActiveMQProducer {
        return ActiveMQProducerImpl(key,vertx,sessionPool.getSession(),destinationType,destination)
    }

}